### PR TITLE
Write the heat equation code with uniform space discretization

### PR DESCRIPTION
**Pull request by thierryantoun**
This PR aims to write the heat equation code with an uniform space discretization.

Changes that were made: 
1.  No more constexpr variables.
2. No more constructor using {}. We use () instead.
3. Comments were removed, so they can be added directly to the associated markdown file in the doc. 
4. Using the DiscreteDomain directly instead of using the [data_x][data_y] signature to extract particular data out of a Chunkspan.

### CODE REVIEW

**Comment by tpadioleau** [comment]
@thierryantoun Please rebase on main

File: examples/uniform_heat_equation.cpp [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    size_t const nb_y_points = 100;
+    double const ky = .002;
+    double const start_time = 0.;
+    double const end_time = 10.;
+    ptrdiff_t const t_output_period = 10;
+    //! [parameters]
+
+    //! [main-start]
+    //! [X-parameters]
+    ddc::DiscreteVector<DDimX> static gwx(1);
```
Comment (tpadioleau) [inline examples/uniform_heat_equation.cpp:107]:
```suggestion
    ddc::DiscreteVector<DDimX> const gwx(1);
```

File: examples/uniform_heat_equation.cpp [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    //! [X-domains]
+    ddc::DiscreteDomain<DDimX> const
+            x_domain_begin(x_domain.front(), x_post_ghost.extents());
+    ddc::DiscreteDomain<DDimX> const x_domain_end(
+            x_domain.back() - x_pre_ghost.extents() + 1,
+            x_pre_ghost.extents());
+    //! [X-domains]
+
+    //! [Y-domains]
+    ddc::DiscreteVector<DDimY> static gwy(1);
```
Comment (tpadioleau) [inline examples/uniform_heat_equation.cpp:128]:
```suggestion
    ddc::DiscreteVector<DDimY> const gwy(1);
```

File: examples/uniform_heat_equation.cpp [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+        //! [numerical scheme]
+        ddc::parallel_for_each(
+                next_temp.domain(),
+                KOKKOS_LAMBDA(
+                        ddc::DiscreteElement<DDimX, DDimY> const ixy) {
+                    ddc::DiscreteElement<DDimX> const ix
+                            = ddc::select<DDimX>(ixy);
+                    ddc::DiscreteElement<DDimY> const iy
+                            = ddc::select<DDimY>(ixy);
```
Comment (tpadioleau) [inline examples/uniform_heat_equation.cpp:264]:
```suggestion
                    ddc::DiscreteElement<DDimX> const ix(ixy);
                    ddc::DiscreteElement<DDimY> const iy(ixy);
```

File: examples/uniform_heat_equation.cpp [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+                        DDimX,
+                        DDimY>(x_domain, y_domain)]);
+    }
+    //! [final output]
+
+#ifdef DDC_BUILD_PDI_WRAPPER
+    PDI_finalize();
+    PC_tree_destroy(&pdi_conf);
+#endif
+}
```
Comment (tpadioleau) [inline examples/uniform_heat_equation.cpp:313]:
```suggestion
}

```

File: examples/uniform_heat_equation.cpp [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+    PDI_init(pdi_conf);
+#endif
+    Kokkos::ScopeGuard const kokkos_scope(argc, argv);
+    ddc::ScopeGuard const ddc_scope(argc, argv);
+
+
+    //! [parameters]
+    double const x_start = -1.;
+    double const x_end = 1.;
+    size_t const nb_x_points = 10;
```
Comment (tpadioleau) [inline examples/uniform_heat_equation.cpp:94]:
```suggestion
    std::size_t const nb_x_points = 10;
```

File: examples/uniform_heat_equation.cpp [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+    double const kx = .01;
+    double const y_start = -1.;
+    double const y_end = 1.;
+    size_t const nb_y_points = 100;
```
Comment (tpadioleau) [inline examples/uniform_heat_equation.cpp:98]:
```suggestion
    std::size_t const nb_y_points = 100;
```

File: examples/uniform_heat_equation.cpp [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+    double const ky = .002;
+    double const start_time = 0.;
+    double const end_time = 10.;
+    ptrdiff_t const t_output_period = 10;
```
Comment (tpadioleau) [inline examples/uniform_heat_equation.cpp:102]:
```suggestion
    std::ptrdiff_t const t_output_period = 10;
```

File: examples/uniform_heat_equation.cpp [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+    //! [initial-conditions]
+    ddc::ChunkSpan const ghosted_initial_temp
+            = ghosted_last_temp.span_view();
+    ddc::parallel_for_each(
+            ddc::DiscreteDomain<DDimX, DDimY>(x_domain, y_domain),
+            KOKKOS_LAMBDA(ddc::DiscreteElement<DDimX, DDimY> const ixy) {
+                double const x
+                        = ddc::coordinate(ddc::select<DDimX>(ixy));
+                double const y
+                        = ddc::coordinate(ddc::select<DDimY>(ixy));
```
Comment (tpadioleau) [inline examples/uniform_heat_equation.cpp:190]:
```suggestion
                double const x
                        = ddc::coordinate(ddc::DiscreteElement<DDimX>(ixy));
                double const y
                        = ddc::coordinate(ddc::DiscreteElement<DDimY>(ixy));
```

File: examples/CMakeLists.txt [hunk 9]
```diff
@@ -4,8 +4,13 @@
 
 add_executable(heat_equation heat_equation.cpp)
 target_link_libraries(heat_equation PUBLIC DDC::DDC)
+
+add_executable(uniform_heat_equation uniform_heat_equation.cpp)
+target_link_libraries(heat_equation PUBLIC DDC::DDC)
```
Comment (tpadioleau) [inline examples/CMakeLists.txt:9]:
```suggestion
target_link_libraries(uniform_heat_equation PUBLIC DDC::DDC)
```

**Review by tpadioleau — CHANGES_REQUESTED** [review]

File: examples/uniform_heat_equation.cpp (hunk 1 above)
Comment (thierryantoun) [reply examples/uniform_heat_equation.cpp:107]: Why was it put in static initially ?

File: examples/uniform_heat_equation.cpp (hunk 3 above)
Comment (thierryantoun) [reply examples/uniform_heat_equation.cpp:264]: `ddc::select` becomes obsolete ?

File: examples/uniform_heat_equation.cpp (hunk 1 above)
Comment (tpadioleau) [reply examples/uniform_heat_equation.cpp]: `static constexpr` ensures that the object is constructed at compile time. `constexpr` only is not enough to ensure it if the object is not used in a constant expression.

File: examples/uniform_heat_equation.cpp [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+    ddc::parallel_deepcopy(ghosted_temp, ghosted_last_temp);
+    display(ddc::coordinate(time_domain.front()),
+            ghosted_temp[x_domain][y_domain]);
+
+    ddc::DiscreteElement<DDimT> last_output_iter = time_domain.front();
+    //! [initial output]
+
+    //! [time iteration]
+    for (auto const iter :
+         time_domain.remove_first(ddc::DiscreteVector<DDimT>(1))) {
```
Comment (tpadioleau) [inline examples/uniform_heat_equation.cpp:214]:
```suggestion
    for (ddc::DiscreteElement<DDimT> const iter :
         time_domain.remove_first(ddc::DiscreteVector<DDimT>(1))) {
```

File: examples/uniform_heat_equation.cpp [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+    double const ky = .002;
+    //! [main-start-y-parameters]
+    //! [main-start-t-parameters]
+    double const start_time = 0.;
+    double const end_time = 10.;
+    //! [main-start-t-parameters]
+    std::ptrdiff_t const t_output_period = 10;
+    //! [parameters]
+
+    //! [main-start]
```
Comment (tpadioleau) [inline examples/uniform_heat_equation.cpp:111]:
```suggestion
    //! [parameters]
    //! [main-start]

```

File: examples/uniform_heat_equation.cpp [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+};
+//! [time-space]
+
+//! [display]
+/** A function to pretty print the temperature
+ * @tparam ChunkType The type of chunk span. This way the template parameters are avoided,
+ *                   should be deduced by the compiler.
+ * @param time The time at which the output is made.
+ * @param temp The temperature at this time-step.
+ */
```
Comment (tpadioleau) [inline examples/uniform_heat_equation.cpp:45]:
```suggestion
/** A function to pretty print the temperature
 * @tparam ChunkType The type of chunk span. This way the template parameters are avoided,
 *                   should be deduced by the compiler.
 * @param time The time at which the output is made.
 * @param temp The temperature at this time-step.
 */
//! [display]
```

File: examples/uniform_heat_equation.cpp [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+    //! [CFL-condition]
+
+    double const dx = ddc::step<DDimX>();
+    double const dy = ddc::step<DDimY>();
+    double const invdx2 = 1. / (dx * dx);
+    double const invdy2 = 1. / (dy * dy);
+
+    ddc::Coordinate<T> const dt(.5 / (kx * invdx2 + ky * invdy2));
+
+    //! [CFL-condition]
```
Comment (tpadioleau) [inline examples/uniform_heat_equation.cpp:160]:
```suggestion
    //! [CFL-condition]
    double const dx = ddc::step<DDimX>();
    double const dy = ddc::step<DDimY>();
    double const invdx2 = 1. / (dx * dx);
    double const invdy2 = 1. / (dy * dy);

    ddc::Coordinate<T> const dt(.5 / (kx * invdx2 + ky * invdy2));
    //! [CFL-condition]
```

File: examples/uniform_heat_equation.cpp [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+    ddc::DiscreteVector<DDimT> const nb_time_steps(
+            std::ceil((end_time - start_time) / dt) + .2);
+
+    ddc::DiscreteDomain<DDimT> const time_domain
+            = ddc::init_discrete_space<DDimT>(DDimT::init<DDimT>(
+                    ddc::Coordinate<T>(start_time),
+                    ddc::Coordinate<T>(end_time),
+                    nb_time_steps + 1));
+
+    //! [time-domain]
```
Comment (tpadioleau) [inline examples/uniform_heat_equation.cpp:172]:
```suggestion
    //! [time-domain]
```

File: docs/first_steps.md [hunk 15]
```diff
@@ -97,9 +92,11 @@ strong typing.
 Instead, for a coordinate in the `X` dimension, we use `Coordinate<X>` and --as already mentioned--
 for a number of elements in the discretization of `X`, we use `DiscreteVector<DDimX>`.
 
-Once this is done, we define two additional domains.
+Once this is done, we define two additional domains:
++ our zone at the start of the domain that will be mirrored to the ghost.
++ our zone at the end of the domain that will be mirrored to the ghost.
```
Comment (tpadioleau) [inline docs/first_steps.md:97]: I guess redundant with the following sentences ?

File: docs/first_steps.md [hunk 16]
```diff
@@ -123,59 +120,96 @@ The first parameter given to the constructor is the first element of the domain,
...
 ## Final swap
 
-\snippet heat_equation.cpp swap
+\snippet uniform_heat_equation.cpp swap
 
 
 # Final output
 
-\snippet heat_equation.cpp final output
+\snippet uniform_heat_equation.cpp final output
```
Comment (tpadioleau) [inline docs/first_steps.md:215]: If we have nothing to say, shall we keep it ?

File: docs/first_steps.md [hunk 17]
```diff
@@ -1,80 +1,75 @@
...
+\snippet uniform_heat_equation.cpp time-space
 
 
 ## Domains
 
 ### Dimension X
 
 Once the types are defined, we can start the `main` function where we will define our various
-domains.
+domains. Here for each dimension, the user needs to specify the starting and ending coordinates of the dimension of the domain, as well as the number of discretization points along each of these dimensions. Additionally, we specify here the physical characteristics specific to our study (the thermal diffusion coefficient).
```
Comment (tpadioleau) [inline docs/first_steps.md:57]:
```suggestion
domains. Here for each dimension, the user needs to specify the starting and ending coordinates of the dimension of the domain, as well as the number of discretization points along each of these dimensions. Additionally, we specify here the physical characteristics specific to our equation (the thermal diffusion coefficient).
```

**Review by tpadioleau — CHANGES_REQUESTED** [review]
Thanks for this contribution, can you also add your name in the AUTHORS file at the root of the repo ?

File: docs/first_steps.md (hunk 16 above)
Comment (thierryantoun) [reply docs/first_steps.md:215]: For me, we should not keep it. I think it adds nothing to the understanding of DDC

File: docs/first_steps.md (hunk 15 above)
Comment (thierryantoun) [reply docs/first_steps.md:97]: Yes, I could have said it in one sentence

Comment (tpadioleau) [reply docs/first_steps.md:97]: I mean the text below explains what the two discrete domains represent right ?

Comment (thierryantoun) [reply docs/first_steps.md]: Yes

**Pull request merged**